server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -49,27 +66,85 @@
                                        final SearchActionListener<SearchPhaseResult> listener) {
+
+        ShardSearchRequest request = rewriteShardSearchRequest(super.buildShardSearchRequest(shardIt));
         getSearchTransport().sendExecuteQuery(getConnection(shardIt.getClusterAlias(), shard.currentNodeId()),
-            buildShardSearchRequest(shardIt), getTask(), listener);
+            request, getTask(), listener);
     }
 
     @Override
     protected void onShardGroupFailure(int shardIndex, Exception exc) {
         progressListener.notifyQueryFailure(shardIndex, exc);
     }
 
+    @Override
+    protected void onShardResult(SearchPhaseResult result, SearchShardIterator shardIt) {
+        try {
+            QuerySearchResult queryResult = result.queryResult();
+            if (queryResult.isNull() == false
+                    && hasPrimaryFieldSort
+                    && queryResult.topDocs().topDocs instanceof TopFieldDocs) {
+                mergeTopDocs((TopFieldDocs) result.queryResult().topDocs().topDocs, result.getShardIndex());
+            }
+        } catch (Exception exc) {
+            getLogger().error("Failed to merge response", exc);
+        }
+        super.onShardResult(result, shardIt);
+    }
+
     @Override
     protected SearchPhase getNextPhase(final SearchPhaseResults<SearchPhaseResult> results, final SearchPhaseContext context) {
         return new FetchSearchPhase(results, searchPhaseController, context);
     }
+
+    // merge the current best bottom field doc with the new query result
+    private synchronized void mergeTopDocs(TopFieldDocs topDocs, int shardIndex) {
+        final ScoreDoc[] bottomDocs;
+        if (topDocs.scoreDocs.length == topDocsSize) {
+            FieldDoc bottom = (FieldDoc) topDocs.scoreDocs[topDocsSize - 1];
+            bottomDocs = new FieldDoc[] { new FieldDoc(bottom.doc, bottom.score, bottom.fields, shardIndex) };
+        } else {
+            bottomDocs = new ScoreDoc[0];
+        }
+        TopFieldDocs toMerge = new TopFieldDocs(topDocs.totalHits, bottomDocs, topDocs.fields);
+        if (bottomTopDocs == null) {
+            bottomTopDocs = toMerge;
+        } else {
+            final Sort sort = new Sort(bottomTopDocs.fields);
+            bottomTopDocs = TopFieldDocs.merge(sort, 0, 1, new TopFieldDocs[]{ bottomTopDocs, toMerge }, false);
+        }
+    }
+
+    ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
+        final TopFieldDocs current = bottomTopDocs;
+        if (current == null) {
+            return request;
+        }
+
+        // disable tracking total hits if we already reached the required estimation.
+        if (trackTotalHitsUpTo != SearchContext.TRACK_TOTAL_HITS_ACCURATE
+                && current.totalHits.value >= trackTotalHitsUpTo
+                && current.totalHits.relation == GREATER_THAN_OR_EQUAL_TO) {
+            assert request.source() != null : "source should contain a primary sort field";
```

Comment:
is this assertion right? I'm not seeing where we make sure to only call this method on sorted requests.